crates/hir_ty/src/infer/expr.rs

```diff
@@ -377,13 +392,13 @@ impl<'a> InferenceContext<'a> {
                     };
 
                 let val_ty = if let Some(expr) = expr {
-                    self.infer_expr(*expr, &Expectation::none())
+                    self.infer_expr(expr, &Expectation::none())
                 } else {
                     TyBuilder::unit()
                 };
 
                 // FIXME: create a synthetic `()` during lowering so we have something to refer to here?
-                let merged_type = self.coerce_merge_branch(*expr, &last_ty, &val_ty);
+                let merged_type = CoerceMany::once(self, last_ty, expr, &val_ty);
```

Comment:
We could store the `CoerceMany` in the breakable context, but then we would eitherh ave to clone the `CoerceMany` out, do the coercion and write it back or run into borrowing issues. Cloning it seems suboptimal if starts to get more fields on it than just the current `Ty`, but then again this current approach wouldn't work either... A bit annoying that the breakable stack is inside the `InferenceContext`

Reply:
Can we `take()` it instead of cloning?

Reply:
You mean via `Option` like this https://github.com/rust-analyzer/rust-analyzer/pull/9512/commits/abbc7e3ed3198529ef390f421a68bc6e128bad92?

Reply:
Yeah. Or alternatively you could `mem::replace` it with a dummy :thinking:

Reply:
I think actually `mem::replace`ing with a dummy might be the best approach. Wouldn't cloning anyhow not work since we want to modify the original one?

Reply:
Oh ye thats what I meant with cloning, clone out and move back in. I suppose `mem::replace` does sound best here.